### PR TITLE
Swift: Add ArithmeticOperation.qll library

### PR DESCRIPTION
**Pull request by geoffw0**
Add an `ArithmeticOperation.qll` library, and make various related changes:
 - add tests for `ArithmeticOperation.qll` and `LogicalOperation.qll`.
 - fix `LogicalOperation.qll` to account for function call `getName` now including the argument pattern.
   - we'd have caught this sooner if there had been a test.
   - also simplified the QL in that library slightly.
 - change `ApplyExpr` to produce a slightly better `toString` in some cases.  I'd really like someone to review this change as I'm not sure how general what I've written is.

I'd also like to know if we're happy with the naming of `BinaryExpr`, `BinaryArithmeticOperation`, `AddExpr` etc, before this stuff becomes more established and difficult to change.

### CODE REVIEW

File: swift/ql/lib/codeql/swift/elements/expr/ApplyExpr.qll
Comment (MathiasVP) [inline swift/ql/lib/codeql/swift/elements/expr/ApplyExpr.qll:15]: If there is no function then there's definitely no static call target. So the `not exists(this.getStaticTarget())` conjunction seems unnecessary.

Really, there shouldn't be any case where where `getFunction` doesn't have a result.

Also, I'm not really a big fan of the strings produced in the example below. The string "call to call to -(_:)" doesn't look like something I'd like to see in an alert message (which is the main point of `toString` in my opinion)

File: swift/ql/lib/codeql/swift/elements/expr/LogicalOperation.qll
Comment (MathiasVP) [inline swift/ql/lib/codeql/swift/elements/expr/LogicalOperation.qll:17]: Thanks for fixing these! I'm surprised it wasn't broken in the CFG tests after I changed the names in https://github.com/github/codeql/pull/9619 🤔.

File: swift/ql/lib/codeql/swift/elements/expr/ArithmeticOperation.qll
Comment (MathiasVP) [inline swift/ql/lib/codeql/swift/elements/expr/ArithmeticOperation.qll:34]: We shouldn't create public abstract AST classes as this creates a dangerous footgun for QL writers. If someone were to create a class that extends `BinaryArithmeticOperation` in a query it'll cause everything that depends on the AST layer to be recomputed.

We either need to make this class `private` (which doesn't seem to be the right thing to do), or list its subclasses in the charpred.

File: swift/ql/lib/codeql/swift/elements/expr/ApplyExpr.qll
Comment (geoffw0) [reply swift/ql/lib/codeql/swift/elements/expr/ApplyExpr.qll:15]: Hmm, I hadn't noticed the repetition of `call to` actually.  I'll see if I can do better.

I agree about `getStaticTarget`.

File: swift/ql/lib/codeql/swift/elements/expr/ArithmeticOperation.qll
Comment (geoffw0) [reply swift/ql/lib/codeql/swift/elements/expr/ArithmeticOperation.qll:34]: Yes, I noticed you'd done something different in `LogicalOperation.qll` and I was meaning to bring up the tradeoff... I don't *like* the way `LogicalAndExpr` doesn't extend `BinaryLogicalOperation`.  I had thought this a problem, but the more I think about it the more I'm not sure it actually is.

I guess I'm talking myself around to adopting your design...

Comment (geoffw0) [reply swift/ql/lib/codeql/swift/elements/expr/ArithmeticOperation.qll]: Done.

File: swift/ql/lib/codeql/swift/elements/expr/ApplyExpr.qll
Comment (geoffw0) [reply swift/ql/lib/codeql/swift/elements/expr/ApplyExpr.qll]: Done the latter, still looking at the former.

File: swift/ql/lib/codeql/swift/elements/expr/ApplyExpr.qll
Comment (MathiasVP) [inline swift/ql/lib/codeql/swift/elements/expr/ApplyExpr.qll:12]: I like the output a lot more now, but I'm not sure we're getting the improved output for the right reasons. It looks like we're now generating `"call to f()"` when we can resolve the call to a static call target `f`, but we print just `"f()"` when we cannot resolve it to a target.

The reason the output looks a lot better now is that the cases where there isn't a static call target are all examples of the form where there's a strange `DotSyntaxCallExpr` (or `SelfApplyExpr`, or a `ConstructorRefCallExpr`) which all have strings that look like `"call to f()"` (because they fall into the category of calls where we can resolve the target).

For example, I think the call to `bar()` in the example below will get a confusing `toString` (since it'll just be the string `"bar"` and not `"call to bar"` even though it's a call to bar and not a reference to the variable).

```swift
func foo(_ bar: () -> ()) {
  bar()
}

foo({ print("I am a lambda :wave:") })
```

Does that make sense?

**Review by MathiasVP — COMMENTED** [review]
I like the new `*Operation.qll` files a lot more now 👍. The only concern I have left is with the `toString` on `ApplyExpr`. If you're okay with it, maybe that should be factored out into a later PR?

**Comment by geoffw0** [comment]
> The only concern I have left is with the toString on ApplyExpr. If you're okay with it, maybe that should be factored out into a later PR?

Yep, that should probably never have been part of this PR.  I've just removed those commits from here (via a rebase / force push - I didn't want to leave a confusing mess here) and will be creating a separate PR for them after this is merged (to avoid conflicts).

File: swift/ql/lib/codeql/swift/elements/expr/ApplyExpr.qll
Comment (geoffw0) [reply swift/ql/lib/codeql/swift/elements/expr/ApplyExpr.qll]: Thanks for this, I need to think about it a bit more.  But since I've factored out this code for a future PR now, it is no longer an issue holding us up here.

**Comment by MathiasVP** [comment]
> Yep, that should probably never have been part of this PR. I've just removed those commits from here (via a rebase / force push - I didn't want to leave a confusing mess here) and will be creating a separate PR for them after this is merged (to avoid conflicts).

That sounds good. In general, I expect the AST structure around function calls to be cleaned up soon, but we're still discussing how to best to this without complicating the extractor too much.

**Comment by MathiasVP** [comment]
I see some of the output in the CFG tests are changing. I'll take a look at it 👀 now!

**Comment by MathiasVP** [comment]
I hope you don't mind that I pushed f35ab7c2924b27137bc12e5f6bf0e98c6bd16533 to your PR. These tests changed because you fixed the names in https://github.com/github/codeql/pull/9737/commits/9e0cf62cdae127b96ef70f49d2e0f56bc1ccc959 which made us generate a better CFG (I apparently broke them in https://github.com/github/codeql/pull/9619 without noticing).

The CFG construction has special handling for `&&` and `||` (which are function calls in Swift), and by fixing the names we now properly identify those function calls again.

**Review by MathiasVP — APPROVED** [review]
LGTM!

**Comment by geoffw0** [comment]
> I hope you don't mind that I pushed https://github.com/github/codeql/commit/f35ab7c2924b27137bc12e5f6bf0e98c6bd16533 to your PR

I don't mind, but in future you probably shouldn't merge your own changes without another set of eyes passing over them.

**Comment by MathiasVP** [comment]
> I don't mind, but in future you probably shouldn't merge your own changes without another set of eyes passing over them.

Good point. I'll keep that in mind!

**Pull request merged**
